### ISSUE TITLE
Sokoban Tandy mode - confusing composite/RGB behavior

### ISSUE THREAD

**Issue by dani-jozsef** [dosbox-staging/dosbox-staging#3452]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### What Operating System are you using?

Linux x86_64

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

When starting Sokoban in Tandy mode, with `composite=off` set in `dosbox-staging.conf`, it looks like this. The relevant log section:
```
2024-02-25 22:42:07.495 | RENDER: Auto-switched to shader 'crt/ega-1440p'
```
```
2024-02-25 22:50:04.203 | DISPLAY: CGA 320x200 16-colour text mode 01h at 59.923 Hz VFR, scaled to 1867x1400 pixels with 1:1.2 (5:6) pixel aspect ratio
2024-02-25 22:50:06.671 | DISPLAY: Tandy 320x200 16-colour graphics mode 09h at 59.923 Hz VFR, scaled to 1867x1400 pixels with 1:1.2 (5:6) pixel aspect ratio
```
![Screenshot from 2024-02-25 22-35-35](https://github.com/dosbox-staging/dosbox-staging/assets/6108136/6d5829a7-fe13-4895-b7c9-115eb3e297e4)

When I switch to `composite = auto` by pressing F12, the screen becomes monochrome(-ish; on the title screen there are some muted blues). I find this inexplicable and strange.
```
2024-02-25 22:43:17.822 | COMPOSITE: State is auto
2024-02-25 22:43:17.870 | DISPLAY: Tandy 640x200 4-colour graphics mode at 59.923 Hz VFR, scaled to 1867x1400 pixels with 1:2.4 (5:12) pixel aspect ratio
```
![Screenshot from 2024-02-25 22-36-03](https://github.com/dosbox-staging/dosbox-staging/assets/6108136/4e22baa4-e938-4999-8acc-3583cc387bab)
![Screenshot from 2024-02-25 23-07-50](https://github.com/dosbox-staging/dosbox-staging/assets/6108136/9b145d74-3aba-483e-a134-17a60a61ccf0)

Pressing F12 again to switch to `composite = on`:
```
2024-02-25 22:45:04.942 | COMPOSITE: State is on
2024-02-25 22:45:04.960 | RENDER: Auto-switched to shader 'crt/composite-1440p'
2024-02-25 22:45:04.960 | VGA: Forcing no pixel doubling
2024-02-25 22:45:04.989 | DISPLAY: Tandy 640x200 composite graphics mode at 59.923 Hz VFR, scaled to 1867x1400 pixels with 1:2.4 (5:12) pixel aspect ratio
```
![Screenshot from 2024-02-25 22-36-11](https://github.com/dosbox-staging/dosbox-staging/assets/6108136/45d438ef-6bcf-402b-b429-585421c605fa)

Now, with F12 I am back to `composite = off`, and I'd expect the original state to be returned. Only it is not.
```
2024-02-25 22:46:29.112 | COMPOSITE: State is off
2024-02-25 22:46:29.121 | RENDER: Auto-switched to shader 'crt/ega-1440p'
2024-02-25 22:46:29.121 | VGA: Pixel doubling enabled
2024-02-25 22:46:29.128 | DISPLAY: Tandy 640x200 4-colour graphics mode at 59.923 Hz VFR, scaled to 1867x1400 pixels with 1:2.4 (5:12) pixel aspect ratio
```
![Screenshot from 2024-02-25 22-36-27](https://github.com/dosbox-staging/dosbox-staging/assets/6108136/f67274a4-1289-4417-8831-dce7bce3dd26)

---

Behavior with other `composite` settings in `dosbox-staging.conf`:

- `composite = auto`: Starts with Tandy color; monochrome in `off` and `auto` modes after switching mode
- `composite = on`: Starts with composite color; monochrome in `off` and `auto` modes after switching mode

---

Behavior is identical in 0.80.1

### Steps to reproduce the behaviour.

1. Run dosbox with `--machine tandy`
2. Launch Sokoban, and select Tandy mode in menu
3. Use `F12` to cycle through composite display behaviors

### Download URL of affected game or software

https://www.abandonwaredos.com/abandonware-game.php?abandonware=Soko-Ban&gid=1239

### Your configuration

```ini
https://gist.github.com/dani-jozsef/b07dfef964e095b7dba8e845c514aaaf
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by johnnovak** [comment]
@dani-jozsef Yeah I noticed this. To be honest, I've spent more time on the Tandy/PCjr code than I ever wanted during my VGA refactors... Ultimately, I don't care about composite mode much as I prefer to play 80s games on the C64 or the Amiga 😄 So this ticket will have to wait for someone with a love for the Tandy/PCjr... As you said, the behaviour is somewhat confusing, but setting composite mode from the config or pressing F12 a couple of times does the job.

**Comment by dani-jozsef** [comment]
At least with IBM CGA cards and early clones, signal would be output simultaneously to both RGBi and composite. The software stack (or the card for that matter) had no idea what monitor you were viewing on. I would assume this was the same on Tandy 1000 machines. I'm quite positive you could literally attach a TV and an RGBi monitor to the same machine, and see the same video output on them simultaneously.

(Hence on many CGA games, you had to manually tell the game whether to optimize graphics for composite color or RGBi...)

So color me quite confused about why this is happening. I remember that dosbox used to treat composite modes as video modes of their own (they technically weren't), somehow auto-detecting it from program behavior, so this behavior could in theory be an artifact of this.

It probably warrants a "test on hardware" tag?

**Comment by johnnovak** [comment]
> At least with IBM CGA cards and early clones, signal would be output simultaneously to both RGBi and composite. The software stack (or the card for that matter) had no idea what monitor you were viewing on. I would assume this was the same on Tandy 1000 machines. I'm quite positive you could literally attach a TV and an RGBi monitor to the same machine, and see the same video output on them simultaneously.

That's good info, thanks, and it makes sense.

> So color me quite confused about why this is happening. I remember that dosbox used to _(ahistorically)_ treat composite modes as video modes of their own (they technically weren't), somehow auto-detecting it from program behavior, so this behavior could in theory be an artifact of this.

Yeah, they are treated as distinct modes. Which is fine—that's only an implementation detail.

See my video mode test wiki page here:
https://github.com/dosbox-staging/dosbox-staging/wiki/Video-emulation-tests#cgapcjrtandy-composite

> It probably warrants a "test on hardware" tag?

Well, I created that tag for myself to test stuff on my real Pentium MMX hardware box so I don't forget about it 😄 So it's meant for internal core team use.

**Comment by Grounded0** [comment]
Dropping https://github.com/dosbox-staging/dosbox-staging/labels/game%20compatibility since the game seems to run properly.
